File: subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        
        if(nums.empty()) return 0;
        
        unordered_map<int,int> map;
        map.insert({0,1});
        
        
        int rSum=0;
        int count=0;
        
        for(int i =0; i< nums.size(); i++){
            rSum=rSum+nums[i];
            int complement = rSum - k;
            if(map.count(complement)){
                
                //auto it=map.find(complement);
                
                 count = count + map.at(complement);
                //count = count + it->second;
                //it->second+1;
            }
            
            if(map.count(rSum)){
                
                // auto it=map.find(rSum);
                // it->second +1 ;
                map.at(rSum)=map.at(rSum)+1;
                
            }else{
                map.insert({rSum,1});
                
            }
                    
        }
        return count;
        
        
        
        
        
        
    }
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        
        hmap = dict()
        hmap[0] = 1

        tsum = 0 
        count = 0

        for i in range(len(nums)):

            tsum = tsum + nums[i]
            compliment = tsum - k


            if compliment in hmap:
                count += hmap[compliment]

       
            if tsum in hmap:
                hmap[tsum] += 1
            else:
                hmap[tsum] = 1
            
            # print (hmap)

        return count
        
        # count = 0
        # for start in range(len(nums)):
        #     for end in range(start + 1, len(nums)+1):
        #         sum_ = sum(nums[start:end])
        #         if sum_ == k:
        #             count += 1
        # return count